In [45]:
import pandas as pd
import numpy as np

# Part I: Train classifiers
### Split data into train set and test set

In [2]:
trainDf=pd.read_pickle('train.pkl')

In [4]:
trainDf.head()

business_id        date               review_id  stars  \
467878   IZI9BdElfGB-WcHpd7Wg7Q  2014-02-20  CPj_ZWalJpnebGrFxV8ujg      5   
450196   Ja4stXdNYr39u5CZHMNtjw  2012-05-14  S1EKSngF9LVvFQ08Pye59Q      4   
1295251  bjllusUsLq9zE4yZnLpOzg  2015-09-19  9ndOEDaRsYjETx0kl83h7w      3   
379392   QdbwZGnBWdkdN2XAG7rwLA  2014-08-15  raKjAf09z4FtAqmqfb5lUw      5   
564809   WKOUTdVJS58E178JjhwidQ  2016-01-10  DQGL9V4vhF7wNBdVRcZ1VQ      4   

                                                      text  \
467878   Came here for a late lunch after being slightl...   
450196   What a cute little place in Old Town!  We went...   
1295251  Food was better than average. We got three dif...   
379392   Korean Style Fried Chicken has become the only...   
564809   Cute little cafe, the crepe was so good. I tri...   

                        user_id  active  
467878   PTcKYk_qU-WzbQjraE3TgQ       1  
450196   l5x4wzz9VOcQ22XMthU6uw       1  
1295251  76crp1vXF1A3LOoNP6zcUQ       1  
379392   4KbUaerUBoiUwZGMHKVFxQ       1  
564809   XrYTMhY9YJvzX2pMepIz7A       1

In [47]:
trainDf.shape

(339700, 7)

In [3]:
X_train=trainDf['text'].values

In [14]:
Y_train=trainDf['active'].values

In [ ]:
testDf=pd.read_pickle('test.pkl')

In [43]:
testDf.head()

business_id        date               review_id  stars  \
840771   LbyCngxEveE7G6w2mqeueg  2015-10-12  PhBoxX0S_htKMGweVmA7iA      4   
524329   5N8R7ALESZ30EoAzVJtabw  2015-11-26  2ZLg2o_yh3pvXnQhdnBCFg      4   
1022586  hPO2di89hgfsD9inbKl7BQ  2014-12-03  z48GmttobtQ28hmKChpSRQ      4   
795885   5B-V6vr6TbtMjrRaBAtHqQ  2012-09-27  Y8jWjRX69baonQSXKLtPHg      2   
769454   u-SJ5QUwrNquL9VnXwl8cg  2013-08-08  UnZirwkho6bJkffEhSWlug      1   

                                                      text  \
840771   Delicious old school pizza, only deliver in a ...   
524329   Probably like a lot of you that reviewed your ...   
1022586  This Ulta has greatly improved! The customer s...   
795885   Not my favorite place to go, but I've only bee...   
769454   I try to like you Postinos, I really do but yo...   

                        user_id  active  
840771   UGlJYHK0I7TM6yLB3Rkp2w       1  
524329   HgCv2lm-GuACRanD7bSqyA       1  
1022586  HV5HZf7DOQHiwCtNdor2Zg       1  
795885   CPodThzsqATG_Bc8a4an_w       1  
769454   k6vOe0_qglL2PZlBldoKeQ       1

In [48]:
testDf.shape

(84926, 7)

In [44]:
X_test=testDf['text'].values
Y_test=testDf['active'].values

### Tokenize traning texts and count word frequency using tf-idf
Reference: http://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html

In [4]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train)
X_train_counts.shape

(339700, 180431)

In [14]:
type(X_train_counts)

scipy.sparse.csr.csr_matrix

In [57]:
tfidf_transformer = TfidfTransformer(sublinear_tf=True)
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
X_train_tfidf.shape

(339700, 180431)

In [49]:
type(X_train_tfidf)

scipy.sparse.csr.csr_matrix

In [58]:
X_test_counts = count_vect.transform(X_test)
X_test_tfidf = tfidf_transformer.transform(X_test_counts)

### Train a naive Bayes classifier and evaluate the performance on the test set

In [59]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train_tfidf, Y_train)

In [60]:
predicted = clf.predict(X_test_tfidf)
np.mean(predicted == Y_test)

0.77591079292560583

We achieve 77.6% accuracy. Not bad!

### Train a linear support vector machine (SVM) model and evalute the performance on the test set

In [61]:
from sklearn.linear_model import SGDClassifier
svm_clf = SGDClassifier(loss='hinge', penalty='l2',alpha=1e-3, random_state=42,max_iter=5, tol=None).fit(X_train_tfidf, Y_train)

In [62]:
svm_clf.coef_

array([[-0.10526299, -0.04814701, -0.00158775, ..., -0.00067528,
        -0.00071631, -0.0008268 ]])

In [63]:
svm_predicted = svm_clf.predict(X_test_tfidf)

In [64]:
np.mean(svm_predicted == Y_test)

0.80533641052210159

In [65]:
from sklearn import metrics
print(metrics.classification_report(Y_test, svm_predicted,target_names=['inactive','active']))

             precision    recall  f1-score   support

   inactive       0.80      0.82      0.81     42811
     active       0.81      0.79      0.80     42115

avg / total       0.81      0.81      0.81     84926



#### SVM is able to predict whether a use in the test set is active by analyzing the contents of the review with an accuracy of 80%.

### Parameter tuning using grid search

In [66]:
from sklearn.model_selection import GridSearchCV
parameters = {'alpha': (1e-2, 1e-3)}
gs_clf = GridSearchCV(svm_clf, parameters)

In [67]:
gs_clf.fit(X_train_counts, Y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=SGDClassifier(alpha=0.001, average=False, class_weight=None, epsilon=0.1,
       eta0=0.0, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='hinge', max_iter=5, n_iter=None,
       n_jobs=1, penalty='l2', power_t=0.5, random_state=42, shuffle=True,
       tol=None, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'alpha': (0.01, 0.001)}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [68]:
gs_clf.best_params_

{'alpha': 0.01}

In [69]:
gs_clf.best_score_

0.826264350897851

In [70]:
svm_clf_tuned = SGDClassifier(loss='hinge', penalty='l2',alpha=0.01, random_state=42,max_iter=5, tol=None).fit(X_train_tfidf, Y_train)

In [71]:
svm_predicted_tuned = svm_clf_tuned.predict(X_test_tfidf)

In [72]:
np.mean(svm_predicted_tuned == Y_test)

0.72279396180203948

## Part II: Visualize & Analyze the result
Reference: https://buhrmann.github.io/tfidf-analysis.html

In [73]:
#holds a list of all the words in the tf-idf’s vocabulary
features=count_vect.get_feature_names()
features

['00',
 '000',
 '0000',
 '00000',
 '0000000',
 '0000000000000001',
 '0000000003',
 '00001',
 '00007651',
 '0002',
 '0002136',
 '0003s',
 '0004318',
 '0004318zz',
 '000dollars',
 '000hr',
 '000iu',
 '000km',
 '000kms',
 '000lbs',
 '000mg',
 '000miles',
 '000s',
 '000sf',
 '000x',
 '001',
 '0015',
 '002',
 '003',
 '0030',
 '004',
 '005',
 '0051',
 '006',
 '0067',
 '007',
 '00719',
 '00789',
 '0089',
 '009',
 '00____',
 '00_____',
 '00______',
 '00a',
 '00am',
 '00dlls',
 '00dollars',
 '00each',
 '00eur',
 '00for',
 '00h',
 '00hh',
 '00hr',
 '00hrs',
 '00ish',
 '00lb',
 '00noon',
 '00off',
 '00or',
 '00p',
 '00pm',
 '00pmish',
 '00s',
 '00tip',
 '00uhr',
 '01',
 '010',
 '0105',
 '0109',
 '011856',
 '012',
 '0131',
 '0135',
 '013672',
 '014',
 '0143',
 '016',
 '017',
 '0176',
 '019',
 '01am',
 '01pm',
 '02',
 '0200',
 '0213',
 '0214',
 '0222000',
 '0233',
 '025',
 '0279',
 '0295',
 '02am',
 '02pm',
 '03',
 '030',
 '0300',
 '033',
 '0330',
 '0359',
 '03607',
 '0362',
 '0368',
 '0383',
 '03a

Create a function that takes a single row of the tf-idf matrix (corresponding to a particular document), and return the n highest scoring words (or more generally tokens or features):

In [74]:
def top_tfidf_feats(row, features, top_n=25):
    ''' Get top n tfidf values in row and return them with their corresponding feature names.'''
    topn_ids = np.argsort(row)[::-1][:top_n]
    top_feats = [(features[i], row[i]) for i in topn_ids]
    df = pd.DataFrame(top_feats)
    df.columns = ['feature', 'tfidf']
    return df

The result of a tf-idf is a sparse matrix, which doesn’t support all the usual matrix or array operations. So in order to apply the above function to inspect a particular document, we convert a single row into dense format first:

In [75]:
def top_feats_in_doc(Xtr, features, row_id, top_n=25):
    ''' Top tfidf features in specific document (matrix row) '''
    row = np.squeeze(Xtr[row_id].toarray())
    return top_tfidf_feats(row, features, top_n)

In [78]:
len(trainDf.loc[trainDf['active']==1])trainDf.iloc[168456]

168457

In [79]:
trainDf.iloc[168456]

business_id                               xmARZ_MuSfrQCXP4e8RQjg
date                                                  2013-03-03
review_id                                 DbTzR3TDySXvYjZRWCO10g
stars                                                          3
text           I want to be able to give Nello's 3.5 stars. \...
user_id                                   BuwzRmdVk0uu17qG49l37A
active                                                         1
Name: 291321, dtype: object

In [80]:
trainDf.iloc[168457]

business_id                               Eif660dJb3nqORM6LF3UYg
date                                                  2015-12-07
review_id                                 a6wkURhUcsDihsNqv0p8LQ
stars                                                          1
text           I was forced by the app to give it a star. Jus...
user_id                                   m44kwkYYFTOMpQM1DeIg4w
active                                                         0
Name: 52298, dtype: object

Due to the way I constructed the train data, *the active users are stored in the first 168457 rows in the trainDf, and the rows are all reviews that belong to inactive users.*

Try using top_feats_in_doc to find the top 10 most important words for the active user on row 0, and those for the active user on row 4.

In [81]:
top_feats_in_doc(X_train_tfidf,features,0,10)

feature     tfidf
0       yellow  0.230024
1         dahl  0.221549
2         whey  0.203711
3        curry  0.189788
4     hungover  0.179498
5       herbal  0.172362
6      sautéed  0.163888
7       thinly  0.159444
8  scrumptious  0.155432
9       triple  0.151808

In [84]:
top_feats_in_doc(X_train_tfidf,features,4,10)

feature     tfidf
0     crepe  0.506723
1  aromatic  0.365277
2     sweet  0.291273
3     latte  0.275677
4    overly  0.252598
5      cafe  0.233309
6    potato  0.224764
7      cute  0.221078
8      okay  0.203687
9       big  0.163168

In [34]:
def top_mean_feats(Xtr, features, grp_ids, min_tfidf=0.1, top_n=25):
    ''' Return the top n features that on average are most important amongst documents in rows
        indentified by indices in grp_ids. '''
    D = Xtr[grp_ids].toarray()
    D = Xtr.toarray()

    D[D < min_tfidf] = 0
    tfidf_means = np.mean(D, axis=0)
    return top_tfidf_feats(tfidf_means, features, top_n)

It would be nice if I can have the top features for all active users, and all inactive users. However, due to the limitation of my laptop's memory, I can only run top_mean_feats for a subset of the users. 

In [ ]:
top_mean_feats(X_train_tfidf,features, slice(0,5000))

In [7]:
D=X_train_tfidf[:5000].toarray()

In [8]:
D[D <= 0.1] = 0

In [9]:
tfidf_means = np.mean(D, axis=0)

In [23]:
len(tfidf_means)

180431

In [18]:
top_tfidf_feats(tfidf_means, features, 50)

feature     tfidf
0        the  0.117207
1        and  0.055301
2        was  0.037273
3         to  0.032746
4         it  0.030709
5         is  0.022037
6         of  0.020370
7        you  0.019955
8         we  0.019106
9       they  0.012762
10      that  0.010780
11       for  0.010238
12       are  0.009502
13      good  0.009110
14        in  0.008685
15      were  0.008405
16       but  0.008248
17      food  0.008225
18     place  0.008070
19   chicken  0.008029
20     their  0.007920
21        as  0.007826
22        my  0.007750
23     pizza  0.007683
24      with  0.007297
25    always  0.006917
26      this  0.006879
27      have  0.006694
28     great  0.006569
29     lunch  0.006220
30    burger  0.006115
31      here  0.006107
32     sushi  0.005997
33       she  0.005902
34    really  0.005901
35       had  0.005649
36      very  0.005576
37        me  0.005523
38        so  0.005494
39  location  0.005321
40     there  0.005226
41      love  0.005194
42        on  0.005168
43  sandwich  0.005008
44      nice  0.004999
45        he  0.004939
46       our  0.004892
47    pretty  0.004851
48     fries  0.004825
49       not  0.004808

In [16]:
D2=X_train_tfidf[-5000:].toarray()

In [17]:
D2[D2 <= 0.1] = 0

In [19]:
tfidf_means2 = np.mean(D2, axis=0)

In [20]:
top_tfidf_feats(tfidf_means2, features, 50)

feature     tfidf
0          the  0.068166
1          and  0.047413
2           to  0.030508
3          was  0.029605
4           we  0.023821
5        great  0.016369
6          she  0.014331
7           he  0.014290
8           my  0.014239
9         food  0.013874
10          is  0.013695
11        they  0.013069
12        very  0.011670
13          it  0.011507
14         our  0.011236
15     service  0.011234
16         you  0.011024
17        best  0.009290
18          of  0.009208
19       place  0.008803
20     amazing  0.008758
21         her  0.008608
22       staff  0.008491
23          me  0.008156
24    friendly  0.008046
25         for  0.008035
26        were  0.008009
27   recommend  0.007712
28        that  0.007487
29          us  0.007416
30        love  0.007364
31         not  0.007310
32        good  0.007200
33         had  0.007089
34     awesome  0.007036
35          in  0.006943
36      highly  0.006882
37  definitely  0.006796
38         are  0.006564
39         car  0.006537
40      always  0.006537
41        will  0.006530
42        hair  0.006492
43    customer  0.006462
44   excellent  0.006373
45        this  0.006336
46       pizza  0.006171
47        have  0.006061
48        ever  0.005877
49          as  0.005839

In [22]:
D3=X_train_tfidf2[-5000:].toarray()

In [23]:
D3[D3 <= 0.1] = 0

In [24]:
tfidf_means3 = np.mean(D3, axis=0)

In [25]:
top_tfidf_feats(tfidf_means3, features, 50)

feature     tfidf
0          great  0.014699
1             we  0.013122
2           food  0.012076
3            she  0.010914
4             he  0.010662
5        service  0.010655
6            was  0.010122
7           very  0.009971
8           best  0.009592
9        amazing  0.009336
10           the  0.009324
11         staff  0.008718
12      friendly  0.008668
13           our  0.008530
14     recommend  0.008427
15        highly  0.007643
16         place  0.007634
17          love  0.007579
18           her  0.007570
19       awesome  0.007326
20    definitely  0.007274
21      customer  0.006783
22            is  0.006684
23     excellent  0.006578
24          good  0.006522
25            us  0.006400
26          will  0.006385
27        always  0.006359
28          ever  0.006271
29  professional  0.006256
30           job  0.006079
31           you  0.005937
32         again  0.005761
33           and  0.005751
34          hair  0.005642
35           car  0.005579
36          nice  0.005578
37         pizza  0.005568
38          were  0.005559
39          back  0.005508
40          they  0.005432
41           his  0.005342
42     delicious  0.005338
43       helpful  0.005149
44          work  0.005121
45          time  0.005087
46        really  0.005038
47         vegas  0.004990
48    experience  0.004982
49    restaurant  0.004978

In [26]:
D4=X_train_tfidf2[:5000].toarray()

In [27]:
D4[D4 <= 0.1] = 0

In [28]:
tfidf_means4 = np.mean(D4, axis=0)

In [29]:
top_tfidf_feats(tfidf_means4, features, 50)

feature     tfidf
0     chicken  0.006611
1          we  0.006401
2      always  0.006118
3       pizza  0.005761
4       lunch  0.005708
5         was  0.005382
6    location  0.005242
7        good  0.005207
8       sushi  0.005145
9      burger  0.005068
10       food  0.004864
11      great  0.004844
12       love  0.004711
13      their  0.004654
14   sandwich  0.004604
15         is  0.004597
16      fries  0.004576
17        you  0.004540
18       soup  0.004366
19       nice  0.004350
20     really  0.004262
21     pretty  0.004240
22      place  0.004232
23      store  0.004220
24        the  0.004203
25       here  0.004095
26        bar  0.004085
27       rice  0.004006
28     coffee  0.003997
29  breakfast  0.003914
30        are  0.003902
31      fresh  0.003895
32  selection  0.003763
33  delicious  0.003733
34     little  0.003711
35         as  0.003688
36        its  0.003675
37      salad  0.003669
38       fish  0.003661
39       pork  0.003653
40        she  0.003648
41     cheese  0.003570
42       fast  0.003564
43        tea  0.003557
44       meat  0.003550
45        ice  0.003472
46     prices  0.003426
47       very  0.003422
48      spicy  0.003401
49       beer  0.003393